### PR TITLE
[READY] Gehennites [Moths Still In Version]

### PR DESCRIPTION
**Pull request by Militaires**
Second iteration, with significant changes.
image
image

- Race of echolocation-using posthumans.
- **Because of intense genetic aberration, they can endure the rigours of spacewalking without any external help.** Low pressure and cold temperature, both are irrelevant to the perfect ship-building engineers.
- They are weaker to fire though
- Almost helpless in combat because of echolocation providing a single frame of events and not a video, and being put on cooldown.
- Their legs are digitigrade and thus they can not wear shoes, their footsteps go tapu tapu though.

Now gehennites can echolocate automatically
Now gehennites can see each other's echolocations, this means gehennites can see better if they're with each other.

I've cut out Gene Liberation, that will come in a different PR. Moths are still in, **merging of the PR or not is still completely up to oranges/maintainers**

:cl: Basilman and Razharas
add:  a new race has been added, Gehennites.
code: Digitigrade limbs can now vary between species
/:cl:

In the future, I intend to give gehennites customization features.

### CODE REVIEW

**Comment by Altoids1** [comment]
Regardless of if this gets merged, I do absolutely love the idea. Echolocation looks absolutely fantastic. I would love to have it in my codebase.

**Comment by TetraK1** [comment]
Echolocation good gehennites bad
why is spaceproof the replacement for gene liberation

**Comment by Militaires** [comment]
image
There were a lot of likes but also a lot of people complaining about 
1. moths being removed 
- [x] 
2. gehennites being just weaker humans on all accounts instead of having something going for them 
- [x]

**Comment by loserXD** [comment]
you forgot to change the changelog

**Comment by Militaires** [comment]
woobs

**Comment by Ratvar** [comment]
Perfect ship-building engineers? Aka cannot wear magboots?

**Comment by Militaires** [comment]
no need for magboots in vacuum

**Comment by Tlaltecuhtli** [comment]
space proof is a boring trait, how about radar like ability to detect moving things out of "vision field"

**Comment by MrFluffster** [comment]
This version is much better, I like how they are now. I hope that being space-capable won't be too much of a buff though.

**Comment by HomocidlePancakes** [comment]
The screenshot shows them all having the same bodies. Is there no cosmetic customization at all?

**Comment by XDTM** [comment]
With how echolocation is coded, could they potentially detect 0 alpha objects like camo'd people?

**Comment by nemvar** [comment]
The perfect space engineer.

> * Race of echolocation-using posthumans.
> * **Because of intense genetic aberration, they can endure the rigours of spacewalking without any external help.** Low pressure and cold temperature, both are irrelevant to the perfect ship-building engineers.

Just one question. How the frick does echolocation work without air?

**Comment by Borisvanmemes** [comment]
Are they fully space proof or just very very space resistent?

**Comment by nero1024** [comment]
@nemvar 
>How the frick does echolocation work without air?

By emitting/radiating some kind microwave beams that bounce off of obstacles? Kind of like radar.

**Comment by ExcessiveUseOfCobblestone** [comment]
was this preapproved (as in with no removal of moths as a stipulation) by someone?

**Comment by anconfuzedrock** [comment]
Why would they need cosmetic customization? They’re fucking blind.

**Comment by Shapsy** [comment]
Aw, I thought the gene suppression lore was actually pretty fuckin tite, kind of a shame to see it just get changed to generic "posthumans"

**Comment by knacker48** [comment]
@anconfuzedrock So you would be ok with naked blind dudes walking around?

**Comment by MrPerson** [comment]
I'm about 99% sure that everyone is going to hate manually echolocating. It really, really should be automatic imo.

**Comment by Repukan** [comment]
If they hate it they can just not play the race

**Comment by Militaires** [comment]
@Shapsy Do not fear my brother, Gene liberated form will still come, I just split it in a different PR so that the controversial part does not sink my whole ship, if people still dislike that after I rework it and it's closed then so be it 

@ExcessiveUseOfCobblestone yes

@nemvar hacks.

**Comment by MrPerson** [comment]
@Repukan That's a dumb argument since it would be funner if it was automatic. I'm applying the same logic we learned from mining scanners basically.

**Comment by datorangebottle** [comment]
This is much better. The spaceproof is a neat little upside to balance out literally being unable to defend yourself, and not super relevant unless you like space exploration- which is hard because you can't fucking see anything. Can they still use consoles / items in hand without being able to see?

I hope the cooldown on their sight isn't incredibly long- i'd say 5, 10 seconds at most; that'd get super annoying super quickly having to push a button every few seconds. It should be more like the auto mining scanner, showing you in slow pulses, imo. A being that lives on echolocation wouldn't have to 'see' manually, they'd just do it, you'd think. Like breathing.

This is a much better PR than the original and I commend you for listening to the comments despite how hostile some of them(and some of mine, sorry) were.

**Comment by Militaires** [comment]
@datorangebottle it's ok bro, yes they can interact with the world normally 100%

@MrPerson i'll PR automated echolocation on its own and if people like it, it can be merged :ok_hand:

**Comment by anconfuzedrock** [comment]
If echo had a hotkey it'd probably be great, but it's odd when you realise nothing special has a hotkey like that. Would it being an outlier be a mistake?

**Comment by anconfuzedrock** [comment]
also, is their echolocation tied to their species, or their organs? it'd be neat if they were missing the eye organ slot and had a special echo tongue/ears, but that's not important immediately or anything.

**Comment by ATH1909** [comment]
Their echolocation supposedly being sound-based (as evidenced by their sensitivity to sounds/their echolocation picking up the bangs from flashbangs) but still working in space is kind of a big issue. If you do want to make it radiation-based or something, make them near-blind near (but able to see) sites of massive radiation but make them not sensitive to sounds/flashbangs.

Will they still have to breathe?

**Comment by Militaires** [comment]
@anconfuzedrock it's tied to their ears, they have eyes but they're permanently blind, i can rip them out at some point i guess, i can add an action button that toggles automated echolocation clicks in the next pr.

@ATH1909 technically they already react to radiowaves, see SM
image
but it's not that big of a deal anyways, and no they don't need to breath, they're spaceproof without any external help.

**Comment by RaveRadbury** [comment]
Echolocation for each footstep. Put those taps to use.

**Comment by zxaber** [comment]
>posthumans

Just call them aliens already. It's nore interesting if they're aliens.
Otherwise, looks good. I'd still recommend making ehcolocation automatic. Having to click a hotbar button constantly _will_ get old quick.

**Comment by OverlyGeneric** [comment]
They look like some form of weird robot or even a troglodyte xenomorph. The lattermost more likely.

Also where floating brain people when? :moyai:

image

**Comment by KomradeSpectre** [comment]
If these are added, I'll probably main one. The race idea is unique, and the fact they're semi-blind is even better.

**Comment by ATH1909** [comment]
> but it's not that big of a deal anyways

But it's ruining my immeeeeeeeeeeeerssiiiooonnnn! Seriously, there really should be at least _some_ justification for them being able to hear/use echolocation in space/vacuums, even if it's just some technobabble that involves bluespace, magic, psionics, or all of the above.

Anyway, +1 to adding automatic pings.

**Comment by Militaires** [comment]
well we can easily handwave it with: they echo psi waves or some bullshit. lore and realism is as irrelevant to authority as it is to me.

**Comment by Militaires** [comment]
###Now gehennites can echolocate automatically (togglable)
###Now gehennites can see each other's echolocations, this means gehennites can see better if they're with each other.**

**Comment by Shadowflame909** [comment]
I wonder how echolocation interacts with things like the morph (which is currently broken and no one seems to know or talk about it) and the chameleon projector

**Comment by MrPerson** [comment]
@Shadowflame909 Like with 0 alpha stuff, it's not too hard to fix issues like that when they crop up.

Make sure to profile this code after it gets merged. Maybe my concerns are overblown, but icon procs are notoriously expensive. I'm ok with echolocation being -kinda- slow though, it's got a lot of cool factor.

As for the flavor of echolocation in space, just say that they're releasing a small burst of x-ray radiation that probably doesn't cause cancer from long-term exposure, according to Nanotrasen in "Employee's Guide to Your Coworkers: How to be Safe and Avoid Lawsuits 5th Edition".

**Comment by Militaires** [comment]
i've cached and then some, in fact, this could EASILY be made faster by dropping special smoothing for walls and dropping unique icons and replacing them with standard overrides, right now, on a test server, echolocation is instant with no lag though so i didnt do that.

**Comment by Shadowflame909** [comment]
Well, even if echolocation can see invisible people. I think that'd be kind of cool, and would really make them different from humans and fit the whole "guard dog" narrative.


I'd nerf the cost of all cloaking items by 1-2 tc though, if that would be the case!

**Comment by Militaires** [comment]
Echolocation somewhat amplifies visibility on low-alpha people, here's the reaction to a 10 alpha human 
image

As long as their alpha isnt exactly ZERO, they're more visible to gehennites, I should probably give them full blown outlines at some point though.

**Comment by anconfuzedrock** [comment]
what if someone had eyes and echo ears?

**Comment by ATH1909** [comment]
> As for the flavor of echolocation in space, just say that they're releasing a small burst of x-ray radiation that probably doesn't cause cancer from long-term exposure, according to Nanotrasen in "Employee's Guide to Your Coworkers: How to be Safe and Avoid Lawsuits 5th Edition".

That's the "make the echolocation radiation-based and not sound-based" idea that was proposed and rejected earlier.

**Comment by Tlaltecuhtli** [comment]
@Militaires how are they able to use machinery like computers?

**Comment by Militaires** [comment]
They can log into camera consoles but won't be able to see anything, they can interact with consoles, but what you should be asking is "can blind people interact with consoles" and not "can gehennites interact with consoles"

**Comment by Mickyan** [comment]
Are you going to remove the blind quirk since it's 4 free points and pretty much obsolete

**Comment by Militaires** [comment]
you're right i should do that, i'll add a blacklist for quirk vs species while im at it

**Comment by SouDescolado** [comment]
About the concerns of Gehennites seeing invisible people, its ok unless they can identify the invisible dude is in fact invisible. AFAIK, they would think it is a normal visible person and would not trigger any alarms.

Also, I see no wires. Can they see wires normally? I would totally roll a Gehennite to go around wiring the solars without the need of a hardsuit.

Can they still see temperature? If so, can they see lamps when they are on as a heat source? An engi Gehe, after wiring the solars, could easily go around the station changing the broken lamps.

If you want Gehe to use computers/cameras, add some kind of pen to them. Imagine is like one of those  Avatar things where they use to ~fuck~ connect to the animals. Handwave lore that the genetic modifications add some kind of uplink, but they need the pen in active hand to be able to use it (It may be needed to refactor some things to make it check if it has pen and not empty hand)

**Comment by anconfuzedrock** [comment]
also can they examine I assume so but

I feel like removing the blind quirk because it'd be "obsolete" is like saying that the civillian class is basically in tf2, because in tf1 the civillian was unarmed with no guns and in tf2 a demoknight doesn't have a gun. Yes, the civilian had only an umbrella to defend himself, yes, the demoman only has a fucking claymore to defend himself, but they aren't really comparable.
Blindness being a ridiculous gimp as a stupid meme is something I'd like to keep around, though of course gehennites shouldn't be able to take it and It's entirely your decision.

**Comment by Militaires** [comment]
@SouDescolado 
yes they can see wires provided the wire isnt hidden by a turf, also the pen sounds like an overly complex feature that boils down to them clicking on a console and using it to me, sorry. i'll make lights unique so they can sense the heat

@anconfuzedrock i assumed he meant not making gehennites be able to take it, removing blindness because of this would be like removing muteness because of mimes, also no they can't examine things because they're blind.

**Comment by SouDescolado** [comment]
@Militaires No need to say sorry to anything. I am just saying things that I thought may have gone unnoticed. Your PR, your rules. I will probably try to play it sometimes, but not always.

**Comment by FantasticFwoosh** [comment]
Those honestly just look like xenomorph people from citadel and other servers aesthetically, Wjohn will be pleased.

Gen(stealers) or whatever they are called seeing ninja sihoulettes and people underneath boxes and in lockers is a feature. Dummy thicc spacemen alerting the gehennites.

**Comment by Militaires** [comment]
>from citadel

yikes, oh you sweet summer child, who hurt you? it's called being a decent human being, let that fucking sink in. 
image

File: code/datums/components/echolocation.dm [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+	var/mob/H = parent
+	var/mutable_appearance/image_output
+	var/list/filtered = list()
+	var/list/turfs = list()
+	var/list/seen = oview(echo_range, H)
+	var/list/receivers = list()
+	receivers += H
+	for(var/I in seen)
+		var/atom/A = I
+		if(!(A.type in echo_blacklist) && !A.invisibility)
```
Comment (kevinz000) [inline code/datums/components/echolocation.dm]: Typecache?

Comment (Militaires) [reply code/datums/components/echolocation.dm]: echo_blacklist is already a typecache

**Comment by gardenBanter** [comment]
snowflakish, and roundstart spacewalking sounds horrible

**Comment by bruhmomento5** [comment]
first race with significant roundstart changes on /tg/.  strong -1 let the events of the round be defined by what actually occurs in the round, not BEFORE it, like changing races, each spaceman should be equal and set apart by skill + job only

**Comment by Militaires** [comment]
they literally are unable to fight back cause they're blind and their vision is limited to 4 turfs :pensive: so you have to be very unrobust to lose to one or even let one escape you

**Comment by bruhmomento5** [comment]
nice adhominem, you can memorize maps and act proactively with this race using features only you have to change combat, not because of shit like your pre-planning but because your character is intrinsically better in those cases

**Comment by Militaires** [comment]
give an example.................

**Comment by bruhmomento5** [comment]
";traitor escape" at roundstart but when you get to escape instead of chasing into maint he just swooces into space and gets away with a dead body

**Comment by Militaires** [comment]
how did he even manage to kill someone

**Comment by bruhmomento5** [comment]
ebow and esword :pepeGRIM:

**Comment by bruhmomento5** [comment]
and game sense

**Comment by Militaires** [comment]
how did he aim while blind

**Comment by bruhmomento5** [comment]
game sense

**Comment by gardenBanter** [comment]
This race has clear roundstart advantages, even if they're situational. How does this have so many pluses? Sure echolocation is pretty sweet but be reasonable

**Comment by Fikou** [comment]
> This race has clear roundstart advantages, even if they're situational. How does this have so many pluses? Sure echolocation is pretty sweet but be reasonable

>roundstart spaceproof with like 5 disadvantages
thats like the only advantage, and any plasmaman that gets an emergency firesuit can do it

**Comment by bruhmomento5** [comment]
plasmaman doing it too doesn't justify it and neither do the disadvantages

**Comment by Fikou** [comment]
> plasmaman doing it too doesn't justify it and neither do the disadvantages

additional fire damage, very weak in combat, no helpful shoes seem like a good trade for being able to spacewalk, when its something an assistant plasmaman can do roundstart and they are supposed to be the hardmode race, and any other race can do with an eva suit or just a mere firesuit as well if they are prepared

**Comment by Militaires** [comment]
image
we did it reddit.

**Comment by anconfuzedrock** [comment]
anyone who thinks this is slightly powerful is a complete fucking idiot who should try playing this game with blind quirk to see why echolocation won't even slightly numb the pain

Ps @Militaires  blind dudes can't use health analyzers for some reason, the one thing that I might hope you look into

**Comment by RaveRadbury** [comment]
So Gehennites can just immediately leave for space, right? Like, I get that they are gonna have a bad time with combat, but they can just leave and not die, right?

**Comment by anconfuzedrock** [comment]
anyone can leave and not die through a disposal unit too, and anyone with a hardsuit can "leave and not die" running into space isn't the only avenue of space in the game, and yes, they are going to have a shit time with combat. it isn't like there's a huge airlock to the void in every room which gehennites could jump out of. and even then, it's not impossible for humans to put in some chase after someone hops out the airlock (which you could avoid by stopping them from getting out the damn airlock

Let's look this over for combat capabilities
-digitigrade so dies to glass shards or lightbulbs (Found in every fucking wall)
-basically blind echolocation won't do enough for you in combat since you'll have like 0.25fps
-more fire damage
-again basically blind
combat advantages:
+can run out an airlock without specialised equipment if the combat takes place near an airlock and they reach an airlock alive and have internals

hmm, maybe, just maybe, being able to run away from combat in certain situations with relative ease isn't a balance issue.

**Comment by RaveRadbury** [comment]
I was thinking more about just going out into space for exploration purposes and how that might be power-gamey, not so much space evasion.

I'm also in favor of this PR, I just want to hear what people think about it.

**Comment by Tlaltecuhtli** [comment]
to space explore they need a jetpack so they have to break in eva anyway

**Comment by Fikou** [comment]
> I was thinking more about just going out into space for exploration purposes and how that might be power-gamey, not so much space evasion.
> 
> I'm also in favor of this PR, I just want to hear what people think about it.

even if you gather space loot unless you want the omnilathe boards and stuff like that or give the loot to others its gonna be pretty useless with 0.25 fps

**Comment by Militaires** [comment]
rock i think ill have to open a PR fixing various shit with blindness like being able to be flashed etc

**Comment by nemvar** [comment]
Again, being flashable while blind is intended. Otherwise revs LITERALLY have no way of converting people with the blind quirk or this new race.

**Comment by actioninja** [comment]
just make blind not immune to the rev flash with some kind of "it's actually auditory" handwave

**Comment by anconfuzedrock** [comment]
or just kill em :^^^^) revs are murdery enough already I don't know what use they'd have for a fucking blind guy plus """emergent gameplay"""
"well this black guy can't be a rev anyways, so we don't need to waste an implant" (plot twist, black guy is rev head)

I'm not worried about space exploration powergaming since again plasmaman firesuit arguably requires less prep since their plasma tank holds so much, and we gave the fucking librarian an eva suit to fuck off (still mad)

I had something else to say last night I forgot but make no mistake it was logical

**Comment by SouDescolado** [comment]
> librarian

Curator

**Comment by anconfuzedrock** [comment]
not important

**Comment by Shadowflame909** [comment]
I'm still sad that the Vampire Hunter whip, wasn't added to the curator and instead went to the chaplain.

Castlevania Curator is more epic then Castlevania Chaplain

**Comment by pizzatiger** [comment]
Loud noises like guns, talking, bombs and certain melee weapons should cause mini echolocation flashes. The Geherines should also have a 3x3 area of nearly permanent echolocation around them when walking around due to their clicking feet, this 3x3 area will also be visible to other Geherines

**Comment by Militaires** [comment]
I'll open a PR flashproofing blindies, plasmamen have flashproof helmets and nearly no one bothers getting the helmet on and off to convert them, if you don't want to bother changing gehennites' bodies to convert them, just kill em. as for you rock I'll open a PR giving them precise echolocation (examining singular items) and maintainers can decide I guess. although I want to say it doesn't make sense that they'll be able to echolocate text on the screen of a health analyzer, they can use consoles anyhow so whatever I won't argue with that.

@pizzatiger what you just described is sound rings, and that feature is already in the works, it's the time-consuming but menial matter of converting our playsounds to datum outputs.

**Comment by anconfuzedrock** [comment]
health analyzers is just because it's so very important to the job, and I always assumed the health analyzer was beeping out the scan results, not displaying it (though i guess that does kinda make way more sense) but ayy 0.07 mm might be able to look at indentations I dunno

**Comment by RaveRadbury** [comment]
@anconfuzedrock That's probably a case where visible_message instead of audible_message. They both allow for an alternate message for the blind/deaf so it could easily be amended to include "The health analyzer states: ..."

**Comment by Militaires** [comment]
after the testmerge some bugs have been made clear, will fix.

**Comment by Militaires** [comment]
by the way the breakage of hats and masks isnt my fault, here's them on my branch and i didnt pull from master yet 
image

**Comment by anconfuzedrock** [comment]
now I remember what my other concern was- until gehennites gain ability to examine, the empath quirk will be useless on them since it's based on examining. a minor point, but a thing.

**Comment by Militaires** [comment]
ill need to setup a quirk X species blacklist

**Comment by KomradeSpectre** [comment]
Don't know if you fixed this bug already, but Gehennites were unable to interact with anything aside from doors and their own inventory unless they either Alt-Click on a tile, or right click on something and use an option from the menu.

**Comment by anconfuzedrock** [comment]
if it aint intended, then maybe change the way echo works so it doesn't effect the 1 tile range around them or something? I mean the issue was with the overlay covering their ability to click shit if they kept using it.

**Comment by optimumtact** [comment]
gotta make that overlay click through bman

**Comment by Qustinnus** [comment]
this seems a lot stronger than humans, why are they spaceproof? they seem like discount skeletons with the downside of giving you eye strain when playing as them.

**Comment by Qustinnus** [comment]
in that regard though; echo devices that make echos and echo visors/devices that let you see echos as if you we're a Gehennite.

also i think the name sucks but thats just me

**Comment by tralezab** [comment]
Honestly seeing almost invisible people clearly with echolocation is enough upsides for the gehennites, they shouldn't be spaceproof

**Comment by Razharas** [comment]
To make hings interactable just makea special subtype of images ofr echoloc specifically with a var set to the source and all click procs redirecting to same procs of the source
Real easy

**Comment by optimumtact** [comment]
that's disgusting

**Comment by anconfuzedrock** [comment]
lmao invisible people are at minimum uncommon and them being someone worth looking for is even rarer

**Comment by Militaires** [comment]
the issue is with mouse opacity, overlays cant have their own mouse opacity, only the opacity of the object, so the problem here is that i either have to set the actual object's opacity to 0 (will fuck with other players) OR add the echolocation images to the mob's screen (but the echo images will move around with you), the third solution is vis_contents object adding to the turf and setting obj opacity to 0, this is what i will experiment with

@Qustinnus 
>downside of giving you eye strain when playing as them.

t. doesn't know how echolocation works ingame

**Comment by MegaEmpirical** [comment]
The echolocation is really cool. Could it be made to provide two seconds of dynamic/video visual before coming to a stop? Would it be possible to have it trigger rwhen a sound is triggered (like a door opening)? Or be continuous when something like an instrument is being played?

**Comment by Militaires** [comment]
all of the things you mentioned are possible but i dont know how laggy they will be, also, in other news, i've come up with some damned solution for the clicking problem, wait until i can optimize it

**Comment by Militaires** [comment]
I've fixed all the issues that I've been reported.

**Comment by actioninja** [comment]
Alright finally got the chance to play as them.
Feedback:
being able to see moving objects on echolocation just fine is a bit odd, even if intended. Changing this would probably be absolutely brutal to play though.
Not being able to tell where you are on the screen is irritating, easily circumvented by physically marking your screen, and also just doesn't much sense. However you seemed to show up sometimes? Not sure what triggers it, or if that's actually the intended functionality.
Wires and pipes render a bit fucked up.

**Comment by JetBrody** [comment]
Just played it, and rolled changeling Round 107828

Super weird interactions with transforming. You get your sight back, but some things that were echolocated stay echolocated. Additionally you can't examine things, you just get the "can't quite see what's there" message.

Also: when you transform back to your gehennite form you can't use your original echolocate buttons, but instead it spawns you a NEW set of buttons that you have to use.

**Comment by Militaires** [comment]
@actioninja you cant see things moving within individual frames, one frame someones here and another he's there and you can sort of interpolate the middle, this excludes dmi animations.

i know what's causing both of the bugs you mentioned, lemme fix

**Comment by Tlaltecuhtli** [comment]
can you make them able to control the range and freq of the radar pulses? like smaller the ranger lower the cooldown between radar pulses

**Comment by anconfuzedrock** [comment]
still some visual errors- medbay jumpsuit from the front is raised slightly for some odd reason.

**Comment by optimumtact** [comment]
with the number of times this calls getflaticon I don't see the other maintainers having much appetite to have this in the game.

Rumour was a native getflaticon was landing in 513, but until something like that lands I don't think this necessarily would be a good idea

**Comment by Militaires** [comment]
it only runs gfi for things in the unique list, and even then gfi isnt necessary and i can substitute it at the cost of pixel-accurate human images, if you want, ill tell you on discord

**Comment by Militaires** [comment]
image
echolocation has been tweaked out of necessity (i had to do this to fix the interactions bug and other solutions were either costly or hacky), now it doesnt provide a SINGLE frame of events, instead it provides outlined vision with movement for a few seconds before periodically fading to black, i tweaked how long they're blind between echolocation clicks to accommodate for this, i've also added an icon showing the center

**Comment by MegaEmpirical** [comment]
It just dawned on me that they should be blind (deaf) in space.

**Comment by Militaires** [comment]
you're not the first one to have that thought but im not gonna nerf them further for the sake of realism

File: tgstation.dme [hunk 2]
```diff
@@ -2403,7 +2406,6 @@
 #include "code\modules\projectiles\guns\ballistic\rifle.dm"
 #include "code\modules\projectiles\guns\ballistic\shotgun.dm"
 #include "code\modules\projectiles\guns\ballistic\toy.dm"
-#include "code\modules\projectiles\guns\energy\dueling.dm"
```
Comment (4dplanner) [inline tgstation.dme]: ?

Comment (Militaires) [reply tgstation.dme]: yeah for some reason byond keeps unchecking that for me, and this time, i didnt catch it

**Comment by IllotAFV** [comment]
So, Gehennite can be stunned by a flash, is this intended? Also, they are getting hallucinations from the SME, but cant wear mesons.

**Comment by MegaEmpirical** [comment]
While a flash makes no sense, hallucinations caused by radiation do. Though I imagine the form they take is 'unusual' for a blind race.

**Comment by Ziiro** [comment]
I think that the spaceproofness might be worth trading out for seeing through walls with the echo location. Traversing a station I know is hard mode, and I can't imagine trying to get around a new/unknown area.

Still this is a pretty wild idea and well executed, good shit

**Comment by ahkent** [comment]
the SME doesnt provide hallucinations from rads- it's literally "if you have it in your visual field, and its active, you gain the hallucinating trait"

**Comment by MegaEmpirical** [comment]
I'm not saying it's induced by the game's mechanic of radiation but by radiation via logic. Otherwise, how else is it affecting you?

**Comment by ahkent** [comment]
just that the fact it's a visual thing mechanically, and the fact that
visually observing it gives a bit of a warning, makes it (fluff-wise) seem
a purely visual thing. the radiation seems like a peripheral effect, if
anything. but hey, it's a game, you know? it doesnt really matter what the
fluff is, it'll get re-figured to account for any added mechanics.

**Comment by Poojawa** [comment]
> Digitigrade limbs can now vary between species


Time to bring back #40514

 :^)

**Comment by Militaires** [comment]
Added caching for everything, and gutted out getflaticon, and even then it's only run for tables and is then cached right after

**Comment by Militaires** [comment]
oops i forgot to push some major shit hold on though

**Comment by vuonojenmustaturska** [comment]
You shouldn't ever have to use getflaticon for game code, it only exists to ship things like humans with a bazillion overlays into browser windows and everything else is a bug. You absolutely shouldn't need to then feed the /icon you just painstakingly constructed into a mutable_appearance that you never modify. Mutable appearances exist to give you an interface to mutating appearances without the intermediate states creating concrete appearances.

**Comment by Militaires** [comment]
They're in mutable appearance form because I was told to use MAs for datum outputs, I use getflaticon on tables so that I can give them a smoothed look, since I can't use the conventional smoothing system, and I cache the result right after so that it's not expensive

**Comment by Militaires** [comment]
by the way the testmerged version of this PR is not up-to-date

**Comment by Razharas** [comment]
Im not sure about movement, dos with sound output feels like a better way, otherwise their vision starts to be barely different from normal

**Comment by Militaires** [comment]
maybe after we get around to converting our 529 sounds and 96039 playsounds to datum outputs

**Comment by AnturK** [comment]
Ping me when you have solution that does not use getFlat and icon operations in droves.

**Comment by Militaires** [comment]
i already cut down gfi to only tables and then made it cache tho

**Comment by ATH1909** [comment]
Is this still gonna be a thing?

**Comment by wesoda25** [comment]
@Militaires you should try the no moth version again cause they've gotten a lot more annoying

**Comment by MegaEmpirical** [comment]
What happened with moths?

**Comment by actioninja** [comment]
This was closed because of performance reasons, and moths are still in the game.
